Add PHP Singleton pattern #43

michiwo1 opened the pull request:
## Related Issue
#43 

## Description
Added a description of the PHP Singleton pattern to the readme. I also wrote code examples in the php file.

michiwo1 committed:
Add PHP Singleton pattern

Dhuds1 commented on patterns/Creational/singleton/PHP/singleton.php:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  // Method to execute a query
+  public function query($sql)
```
This isn't a safe way to query data, as it can allow SQL injections, use something like

```PHP
public function query($sql, $params = []) {
  $stmt = $this->connection->prepare($sql);
  $stmt->execute($params);
  return $stmt;
}
```

michiwo1 replied:
I've fixed this.
8e2d17072a0391e24f96838aed96aacb8e23c0bb

Dhuds1 commented on patterns/Creational/singleton/PHP/singleton.php:
```diff
@@ -0,0 +1,45 @@
+
+// Usage
+$db = DatabaseConnection::getInstance();
+$result = $db->query("SELECT * FROM users");
```
using the previous code as an example:

```PHP
$stmt = "SELECT * FROM users";
$results = $db->query($stmt);
```

michiwo1 replied:
I've fixed this too.
33143430f1bb7c3a1bd6a9d0c5f6dd10ea37e60a

Dhuds1 reviewed (commented):
Your code looks well structured! I recommend redoing the query function as its an unsafe way to query the DB, and depending on the applications use I would also recommend maybe including a `findOrFail` method.

Lastly, because this is a DB querying class ~kinda, you can consider simplifying the class name to `DB` to make using the class method calls easier with `DB::getInstance()`.

if you were to introduce this class structure, you could also make the query method static and call it like `DB::query();`

michiwo1 committed:
Refactor query method to use prepared statements

michiwo1 committed:
Refactor query execution in singleton.php

michiwo1 committed:
Add findOrFail method

michiwo1 committed:
Refactor DatabaseConnection class to use DB as the class name

michiwo1 commented:
# Update 1
## Details
### I have updated the method to ensure a safer way to query data, thereby mitigating the risk of SQL injections.
- comment:   https://github.com/nic-dgl104-winter-2024/pattern-library/pull/83#discussion_r1538076870
- commit: 8e2d17072a0391e24f96838aed96aacb8e23c0bb
### Refactoring of some code was performed
- comment: https://github.com/nic-dgl104-winter-2024/pattern-library/pull/83#discussion_r1538085252
- comment: https://github.com/nic-dgl104-winter-2024/pattern-library/pull/83#pullrequestreview-1958521398
- commit: 33143430f1bb7c3a1bd6a9d0c5f6dd10ea37e60a
- commit: 15c86a2be067adb877d7fc1579dff24e0689dd89
### Added findOrFail method
- comment: https://github.com/nic-dgl104-winter-2024/pattern-library/pull/83#pullrequestreview-1958521398
- commit: a15f362da06ab553cd3dc3c8229b5a2d3a258ed9

michiwo1 commented:
Hi @Dhuds1,
Thanks for doing your review!
I have fixed some of your points and added some of your ideas as well. I hope you will review it again when you have time.
The following is a comment summarizing what has been fixed.
https://github.com/nic-dgl104-winter-2024/pattern-library/pull/83#issuecomment-2019027613

Dhuds1 commented on patterns/Creational/singleton/PHP/singleton.php:
```diff
@@ -0,0 +1,58 @@
+  {
+  }
+}
+
```
Looks good! you need to add
```PHP
$db = DB::getInstace();
```

michiwo1 replied:
I've fixed it.
f3d398cf45729832f2c47d29a833328f61250722

michiwo1 committed:
Add singleton instance of DB class

Dhuds1 commented:
Looks great! I don't see anything else to comment on!

ash-teach commented:
Thanks @michiwo1 and @Dhuds1!
